### PR TITLE
Add recipe for trimspace-mode

### PR DESCRIPTION
**Pull request by bkhl**
A minor mode to trim trailing whitespace/newlines.

### Brief summary of what the package does

This is a package I have used myself for a couple of years now and think is quite useful.

The main component is a minor mode that simply sets local variables/hooks to enable Emacs' own automatic trimming of trailing space/newlines on save.

It has a function `trimspace-mode-unless-trailing-whitespace` which can be used as a hook for all file modes to enable it if you open a file that already is "clean" of trailing whitespace.

To use it on a file that has trailing whitespace since earlier you just manually enable the mode. 

### Direct link to the package repository

https://git.sr.ht/~bkhl/trimspace-mode/

### Your association with the package

Author and maintainer.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

### CODE REVIEW

**Comment by riscy** [comment]
Thanks for this.  I know there are a few [packages in this "space"](https://melpa.org/#/?q=whitespace) :) can you distinguish what makes yours different?  I'd check out e.g. ws-butler, whitespace-cleanup-mode, smart-hungry-delete, shrink-whitespace, ethan-whitespace.  There's also the builtin whitespace.el which contains `whitespace-cleanup`.

I think distinguishing (or disambiguating?) your package from these - if it makes sense so - could lend itself to really good documentation that would inform your users "here's why this package is different.  Use these packages if you're looking for X..."

Otherwise the code and checklist looks good.

**Comment by bkhl** [comment]
Some of those I remember seeing before I wrote my own mode, some seem to be newer additions.

This first list actually have different functionality:

* `ws-butler`: more complex mode that trims whitespace only on "touched" lines.
* `smart-hungry-delete` / `shrink-whitespace`: these are about editing whitespace in between words &c., so it's not about trailing whitespace.
* `ethan-whitespace` / `whitespace.el`: significantly more complex modes that also does faces for trailing whitespace and stuff.

This one though is quite similar:

* `whitespace-cleanup-mode`

However, I'll stick to my own because:

* instead of reimplementing the "cleanup" functions it just uses built-in variables/hooks for that
* as a consequence it is roughly half the lines of code when ignoring.

Had I been aware of `whitespace-cleanup-mode` a few years ago I might not have written my mode, but as I think I prefer it for those reasons. Up to you if you think those differences are too weak an argument to include it in MELPA.

The documentation already makes clear that this is how it works. I rather not maintain a list of comparisons to other packages there, as I'd feel obliged to keep that up to date.

**Comment by riscy** [comment]
I think there's a big enough overlap that it makes sense not to merge - until maybe sometime in the future should the feature sets really diverge.  Perhaps @purcell would be interested in seeing or adopting your use of built-in cleanup functions, though.

I'll close this for now, but feel free to followup any time. :)

**Comment by bkhl** [comment]
I don't think @purcell's use of `whitespace-mode` rather than `delete-trailing-whitespace` from `simple.el` is necessarily worse, and some of the added features around fixing weird tabs and such in `whitespace-mode` seems nice.

The only thing holding me from just adopting `whitespace-cleanup-mode` right now is the lack of management of `require-final-newline` in `whitespace-mode`.

**Comment by bkhl** [comment]
@riscy I'd like to reopen this actually. I raised an issue over at https://github.com/purcell/whitespace-cleanup-mode/issues/19 but got redirected to Emacs itself, since it turns out `whitespace-cleanup-mode` is trying to stick to using `whitespace.el` which is included in Emacs for these operations.

I made a patch for that which looks like it will get into Emacs 31 (https://lists.gnu.org/archive/html/bug-gnu-emacs/2024-06/msg00875.html).

However, while writing that patch I ran into other bugs like not actually removing all trailing newlines when you configured it to do so. Since all I really wanted was to strip trailing space/newlines, I'm left with no mode other than my own that actually does the same thing, meaning I do think there would be a value in adding it to MELPA after all.

**Comment by riscy** [comment]
Yes, we can reopen this - but I think since you've deleted the branch, only you can reopen it. :)

**Comment by bkhl** [comment]
@riscy I have restored the branch but don't think I'm able to reopen.

**Comment by riscy** [comment]
Interesting that you can't reopen - but now that you've restored the branch I can reopen.

**Comment by riscy** [comment]
I read the exchange in https://github.com/purcell/whitespace-cleanup-mode/issues/19 and I'll try to summarize here:

- whitespace-mode does *nearly* what you want it to, but there is specific behavior around trailing newlines that you want to enforce.  That led to a fix slated for Emacs 31 (congrats on getting that fixed for everyone)
- You noticed several other bugs (not features?) in the builtin whitespace.el; you didn't expand on these but I imagine if you just say what they are, nobody would fault you for not actually fixing them, and you'll increase the likelihood that they'll get some attention

On the one hand, I agree with you that there is value around people using your package who are running into the same bugs you are.  But more generally I think this wouldn't be a good policy for MELPA, i.e. Emacs builtin x has a bug, MELPA introduces package y that does the same thing but with the bug fixed.

I'm still leaning toward it being a good idea for us to *not* merge this, but if you could describe the other bugs it would really help to make an informed decision.

**Comment by bkhl** [comment]
I'll respond in two parts:

The new bug I found was that `whitespace-cleanup` doesn't actually properly clean up empty lines at end of the buffer (a function it already had). It will still leave one empty line. I may actually since I started that thread go ahead and amend my patch to get that fixed though, since I've already taken up the developers time and it seems worthwhile.

However:

While doing this I understood `whitespace-mode` doesn't actually just deal with faulty trailing newlines/whitespace, but also a bunch of other stuff like bad mixes of tab/space etc. I don't feel a need for this, which is why I'll go back to  my own mode which only cares about _trailing_ whitespace.

In other words, when I did my initial list of comparisons, I should have categorized `whitespace-cleanup-mode` in with `smart-hungry-delete` &c. as a much more complex mode, or possible together with `whitespace.el`, as it turns out it's basically an extension to that.

This is why I'll stick to my mode, and think there may be others out there that makes the same judgement. I'll leave it up to you to say if that's sufficient justification to include it in MELPA or not, I'll of course keep it up on the repo and maintain it regardless.

**Comment by riscy** [comment]
Would you be willing to add these distinguishing features to your `;;; Commentary`?  Otherwise I fear this may just get lost in the MELPA index as possibly dupilcate functionality.  But if we can do that, I think you make a strong case.

**Comment by bkhl** [comment]
I will do that, though I'll avoid making a list of comparisons particular other packages, as that would make me feel obliged to keep that updated.

**Comment by riscy** [comment]
Okay, I'll merge this.
